### ISSUE TITLE
`Cannot read property 'prototype' of undefined` error when installed with NPM v3

### ISSUE THREAD

**Issue by Bill77** [ReadyTalk/cukefarm#35]
Hi @thompsnm,

Interesting setup for testing with page objects and cucumber. We are hoping to use the framework, although I'm having trouble with the setup.

When I define the `World.js` and `Protractor.conf.js` as described, I get a undefined access error when I try to access the objects (`World` or `config`). 

It seems I'm creating a circular dependency between World.js and the index.js because 'rek' is trying to find the file that I'm creating. Or it's possible I don't understand NodeJs well enough, and this is a completely separate problem. :)

Any help would be appreciated, TIA,
Bill

``` javascript
// local World.js
var World = require('cukefarm').World;

World.prototype.pageObjectMap = require('./PageObjectMap'); //<-- I have an undefined error on 'prototpye'
module.exports.World = World

// cukeFarm index.js
rek = require('rekuire');
World = rek('World').World; //<-- isn't this the local World.js being created?
config = rek('protractor.conf').config;

module.exports = {
  World: World,
  config: config
}

**Comment by thompsnm** [comment]
Hey @Bill77 ,

Thanks for evaluating CukeFarm, I hope it works out for you! It's tough to say exactly what's going wrong; everything looks fine and I don't immediately recognize that error as something I've encountered before. If you pasted your full stracktrace we might be able dig into the error, but that would probably take a few days of back-and-forth to sort through. Before we jump into something like that, I would recommend that you take a look at this CukeFarm template project I created:

https://github.com/thompsnm/cukefarm-template

It's a fully functional (albeit contrived) test suite using CukeFarm. As long as you're running Node v4.0.0 or greater you should just need to clone the project, run an npm install, then run an npm test to kick off the suite. (If you're using a Node version less than v4.0.0 you can still use the project, just make sure you checkout a commit from before I upgraded to Cukefarm 3.) You can compare that to what you've already written to try to see where the two deviate, or you could even just delete the fake app / example code and use the template as a starting point for your own suite.

Hopefully the template can give you some answers! And let me know if you discover what the issue is so I can try to anticipate that sort of error in the CukeFarm documentation.

**Comment by Bill77** [comment]
Hey @thompsnm,

Thanks for the quick response, I've cloned and tried to execute your template project but unfortunately get the same error. It seems I'm dealing with the same issue (perhaps with my setup?). I've attached the console output below, hoping that will help. Just some quick notes about my setup:
* Win 7 Env -- I know :-)
* Updated NodeJs to v.5.4
* Had installed protractor, cucumber, protractor-cucumber-framework to global (was using gulp with a local versions on my project)

[Console Output](https://github.com/ReadyTalk/cukefarm/files/88108/console.txt)

**Comment by thompsnm** [comment]
Ah, thanks for the additional info! I'm using OSX with Node v4.2.2, so I've got a substantially different environment. Your stacktrace got me wondering if there's an issue with how the rekuire library was loading in modules on your setup, so I did some digging and discovered someone having a similar problem reached out the rekuire guys:

https://github.com/nadav-dav/rekuire/issues/14

There's a lot of info in that thread, so I'll find some time to dig through that later.

**Comment by Bill77** [comment]
Interesting research, I've commented on that thread to hopefully help them resolve the issue. 

Thinking along your lines, I removed references to `rek` in CukeFarm project and now things are working. 

If you'd like me to submit a pull request, I'd be happy to contribute.

**Comment by lukefan6** [comment]
Ha I was actually the one who posted that thread on Rekuire 
At first I suspect it's the cukefarm issue but after some digging I thought it could have something to do with npm versions since npm treat node_modules structures very differently between v2 / v3.
I can get my cukefarm project running on npm v2 pretty well
However I think we could somehow work together and figure it out how to get it working on npm v3 since we don't want our project to be "NPM V2 ONLY", that just feel bad :(
Looking forward to our solutions, cheers guys!

**Comment by thompsnm** [comment]
I'm still having trouble reproducing this issue. I tried cloning cukefarm-template on another machine with the following setup:

* Windows 7 - 64-bit
* NodeJs v5.4.1
* npm 3.3.12

Running the tests worked fine:
[cukefarm-template_test_results_windows.txt](https://github.com/ReadyTalk/cukefarm/files/91170/cukefarm-template_test_results_windows.txt)

What's weird though is that cloning rekuire and running the unit tests on that same machine yielded the same failures that @Bill77 posted in the linked rekuire issue. I'll post my results there to corroborate, but I'll have to do some more testing to reproduce the reported issue with CukeFarm.

**Comment by Bill77** [comment]
@thompsnm 
Hmm, I wish I was getting those same results. :-)

I'm running on exact same setup, but still running into issues with Rekuire not finding the modules properly. Is there something different with your npm global setup? Here is my global package list

```console
c:\elekta>npm -g ls -depth=0
C:\Users\uschebil\AppData\Roaming\npm
├── bower@1.7.1
├── cordova@4.3.0
├── cucumber@0.9.2
├── generator-hottowel@0.3.7
├── grunt@0.4.5
├── grunt-cli@0.1.13
├── gulp@3.8.11
├── gulp-cli@0.3.0
├── ionic@1.3.11
├── normalize-package-data@2.3.5
├── npm@3.3.8
├── npm-windows-upgrade@0.4.0
├── protractor@3.0.0
├── protractor-cucumber-framework@0.3.2
└── yo@1.4.8
```

@lukefan6, perhaps you can be a tie-breaker? Are you able to run [cukefarm-template](https://github.com/thompsnm/cukefarm-template)?

Another interesting thing, I tried forcing rekuire to ignore the 'spec' folder so that it'll attempt to find the local world.js file. And here is what I got, looks like Rekuire isn't able to find the local world.js within cukefarm:

package.json
``` json
{
  "name": "cukefarm-template",
  "version": "1.0.0",
  "description": "Template for creating test suites with CukeFarm",
  "author": "Nathan Thompson",
  "license": "MIT",
  "devDependencies": {
    "cukefarm": "^3.0.0",
    "express": "^4.13.1",
    "globule": "^0.2.0",
    "grunt": "^0.4.5",
    "grunt-cli": "^0.1.13",
    "grunt-contrib-connect": "^0.10.1",
    "grunt-protractor3-runner": "^3.0.0",
    "jit-grunt": "^0.9.0",
    "time-grunt": "^1.0.0"
  },
  "scripts": {
    "test": "webdriver-manager update; node app.js & appId=$!; protractor ./spec/conf/protractor.conf.js; kill $appId",
    "test-grunt": "grunt"
  },
  "rekuire": {
      "ignore": [
          "spec"
      ]
  }
  
}
```

And here is my output:
``` console
Running "protractor:chrome" (protractor) task
webdriver-manager path: c:\GitHub\cukefarm-template\node_modules\protractor\bin\webdriver-manager
selenium standalone is up to date.
chromedriver is up to date.
ERROR - failed loading configuration file ./spec/conf/protractor.conf.js
c:\GitHub\cukefarm-template\node_modules\protractor\lib\configParser.js:184
    throw e;
    ^

Error: Can not find a module by the name of [World] or it has returned empty. nested: Error: Cannot find module 'c:\GitHub\cukefarm-template\node_modules\cukefarm\World'
Error: Cannot find module 'World'

    at getModule (c:\GitHub\cukefarm-template\node_modules\rekuire\lib\main.js:104:15)
    at rekuire (c:\GitHub\cukefarm-template\node_modules\rekuire\lib\main.js:28:12)
    at Object.<anonymous> (c:\GitHub\cukefarm-template\node_modules\cukefarm\index.js:2:9)
    at Module._compile (module.js:397:26)
    at Object.Module._extensions..js (module.js:404:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (c:\GitHub\cukefarm-template\spec\conf\protractor.conf.js:1:76)
```

**Comment by lukefan6** [comment]
ok, I tried to run cukefarm-template on Yosemate with node version v4.2.4
as I suspected on the other thread, I tried with two different npm versions
on npm v2.14.15, the test passed
however on npm v3.5.3, the test failed with following messages:

```
ERROR - failed loading configuration file ./spec/conf/protractor.conf.js
/Users/lukefan/cukefarm-template/node_modules/protractor/lib/configParser.js:184
    throw e;
    ^

TypeError: Cannot read property 'prototype' of undefined
    at Object.<anonymous> (/Users/lukefan/cukefarm-template/spec/support/World.js:5:6)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:313:12)
    at Module.require (module.js:366:17)
    at getModule (/Users/lukefan/cukefarm-template/node_modules/rekuire/lib/main.js:81:22)
    at rekuire (/Users/lukefan/cukefarm-template/node_modules/rekuire/lib/main.js:28:12)
    at Object.<anonymous> (/Users/lukefan/cukefarm-template/node_modules/cukefarm/index.js:2:9)
    at Module._compile (module.js:435:26)
```

it's basically the same situation that @Bill77 mentioned in the very beginning of this thread

and my conclusion remains the same that it has something to do with npm versions because the folder structure inside node_modules are very different between npm v2 and v3 so rekuire was not able to locate modules properly

guys what are your thoughts?

**Comment by thompsnm** [comment]
Ah, I was finally able to reproduce the issue! I must have been messing up the npm install somehow on my earlier attempts.

@Bill77 , you mentioned earlier that you got CukeFarm working by removing references to `rek`, could you submit that pull request now? I'd be interested in looking it over now that I'm sure I can verify the fix on my machine. Also, it'd be nice to remove `rekuire` if it's no longer necessary. The fewer dependencies the better in my opinion.

**Comment by Bill77** [comment]
@thompsnm, sure I'll fork and create a pull request. The only tricky reference is with the generalStepsDef.js which needs to be included in the protractor.conf.js. I just referenced it based on the needed pathing, although I'm open to a better solution. :)

Also I'd like to contribute some documentation and error handling since they were hangups for me when initially using it, so I'll make a separate request then. 

Glad the issue is getting resolved. :)

**Comment by thompsnm** [comment]
This should be fixed in `v3.0.1`
